oxenstored/transaction.ml

```diff
@@ -219,9 +219,76 @@ let rm t perm path =
   add_wop t Xenbus.Xb.Op.Rm path
 
 let ls t perm path =
-  let r = Store.ls t.store perm path in
+  let _node, r = Store.ls t.store perm path in
   set_read_lowpath t path ; r
 
+let ls_partial t perm path directory_cache =
+  let configurable_cache_size = 8 in
+  let check_hashtable_bounds ht =
+    if Hashtbl.length ht >= configurable_cache_size then
+      let key, _min_timestamp =
+        Hashtbl.fold
+          (fun key (_, _, _, timestamp) (min_key, min_timestamp) ->
+            if timestamp < min_timestamp then
+              (Some key, timestamp)
+            else
+              (min_key, min_timestamp)
+          )
+          ht (None, infinity)
+      in
+      Hashtbl.remove ht (Option.get key)
+  in
+
+  (* Return the cached buffer if the node hasn't changed; otherwise
+     create or refresh the cache *)
+  let ( let* ) = Option.bind in
+  let cache_opt = Hashtbl.find_opt directory_cache path in
+  let cache =
+    let* cached_node, gen_count, cache, _timestamp = cache_opt in
+
+    let* cur_node = Store.get_node t.store path in
+    (* Regenerate the cache entry if the underlying node changed *)
+    if cached_node == cur_node then
+      (* TODO check_parents_perms_identical oldroot currentroot path*)
+      Some (cached_node, gen_count, cache)
+    else
+      None
+  in
+  let ret =
+    match cache with
+    | Some (cached_node, gen_count, cache) ->
+        (* Update the timestamp in the cache *)
+        Hashtbl.replace directory_cache path
+          (cached_node, gen_count, cache, Unix.gettimeofday ()) ;
+        (gen_count, cache)
+    | None ->
+        let node, entries = Store.ls t.store perm path in
+        let r =
+          if List.length entries > 0 then
+            Utils.join_by_null entries ^ "\000"
+          else
+            ""
+        in
+        (* We need to increment the generation count if the cache existed
+           but needed to be refreshed *)
+        let generation_count =
+          match cache_opt with
+          | Some (_, generation_count, _, _) ->
+              Int64.add generation_count 1L
+          | None ->
+              1L
```

Comment:
I think we need to remember what the largest previous value was and start there, or some other way to ensure it is unique. Otherwise if we have to evict something from the cache (because the size got exceeded), and we readd it, then it'll go back to 1 and the client may not notice that it has changed.

We could store a highest-evicted-counter, or a counter of how many operations we've done on the connection cache, etc.

Reply:
I can do that, but then if the cache entry was evicted, the client will be forced to restart the sequence of calls as the generation count will have some new value. This is better than being oblivious to changes inbetween calls, of course.